File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters.Remote/Database/DatabaseRecord.cs
// <copyright file="RemoteDatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Json.Api.Sync;
    using Meta;
    using Shared.Ranges;

    internal class DatabaseRecord : Adapters.DatabaseRecord
    {
        private readonly DatabaseConnection database;

        private Dictionary<RelationType, object> roleByRelationType;
        private SyncResponseRole[] syncResponseRoles;

        internal DatabaseRecord(DatabaseConnection database, Class @class, long id, long version) : base(@class, id, version) => this.database = database;

        internal static DatabaseRecord FromResponse(DatabaseConnection database, ResponseContext ctx, SyncResponseObject syncResponseObject) =>
            new DatabaseRecord(database, (Class)database.Configuration.MetaPopulation.FindByTag(syncResponseObject.c), syncResponseObject.i, syncResponseObject.v)
            {
                syncResponseRoles = syncResponseObject.ro,
                GrantIds = ValueRange<long>.Load(ctx.CheckForMissingGrants(syncResponseObject.g)),
                RevocationIds = ValueRange<long>.Load(ctx.CheckForMissingRevocations(syncResponseObject.r))
            };

        internal ValueRange<long> GrantIds { get; private set; }

        internal ValueRange<long> RevocationIds { get; private set; }

        private Dictionary<RelationType, object> RoleByRelationType
        {
            get
            {
                if (this.syncResponseRoles != null)
                {
                    var metaPopulation = this.database.Configuration.MetaPopulation;
                    this.roleByRelationType = this.syncResponseRoles.ToDictionary(
                        v => (RelationType)metaPopulation.FindByTag(v.t),
                        v =>
                        {
                            var roleType = ((RelationType)metaPopulation.FindByTag(v.t)).RoleType;
                            var objectType = roleType.ObjectType;

                            if (objectType.IsUnit)
                            {
                                return this.database.UnitConvert.UnitFromJson(objectType.Tag, v.v);
                            }

                            if (roleType.IsOne)
                            {
                                return v.o;
                            }

                            return ValueRange<long>.Load(v.c);
                        });

                    this.syncResponseRoles = null;
                }

                return this.roleByRelationType;
            }
        }

        public override object GetRole(RoleType roleType)
        {
            object @object = null;
            this.RoleByRelationType?.TryGetValue(roleType.RelationType, out @object);
            return @object;
        }

        public override bool IsPermitted(long permission)
        {
            if (this.GrantIds.IsEmpty)
            {
                return false;
            }

            if (this.RevocationIds.Any(v => this.database.RevocationById[v].PermissionIds.Contains(permission)))
            {
                return false;
            }


            return this.GrantIds.Any(v => this.database.GrantById[v].PermissionIds.Contains(permission));
        }
    }
}

File: typescript/workspace/system/adapters-json/src/lib/database/database-record.ts
import { Class, RoleType } from '@allors/workspace-system-meta';
import {
  DatabaseRecord as SystemDatabaseRecord,
  IRange,
} from '@allors/workspace-system-adapters';
import {
  SyncResponseObject,
  SyncResponseRole,
} from '@allors/database-system-protocol-json';
import { DatabaseConnection } from './database-connection';
import { ResponseContext } from './security/response-context';
import { unitFromJson } from '../json/from-json';

export class DatabaseRecord extends SystemDatabaseRecord {
  grants: IRange<number>;
  revocations: IRange<number>;

  private _roleByRoleType?: Map<RoleType, unknown>;
  private syncResponseRoles?: SyncResponseRole[];

  constructor(
    public readonly database: DatabaseConnection,
    cls: Class,
    id: number,
    public override version: number
  ) {
    super(cls, id, version);
  }

  static fromResponse(
    database: DatabaseConnection,
    ctx: ResponseContext,
    syncResponseObject: SyncResponseObject
  ): DatabaseRecord {
    const object = new DatabaseRecord(
      database,
      database.configuration.metaPopulation.metaObjectByTag.get(
        syncResponseObject.c
      ) as Class,
      syncResponseObject.i,
      syncResponseObject.v
    );
    object.syncResponseRoles = syncResponseObject.ro;
    object.grants = ctx.checkForMissingGrants(syncResponseObject.g);
    object.revocations = ctx.checkForMissingRevocations(syncResponseObject.r);
    return object;
  }

  get roleByRoleType(): Map<RoleType, unknown> {
    if (this.syncResponseRoles != null) {
      const metaPopulation = this.database.configuration.metaPopulation;
      this._roleByRoleType = new Map(
        this.syncResponseRoles.map((v) => {
          const roleType = metaPopulation.metaObjectByTag.get(v.t) as RoleType;
          if (roleType == null) {
            throw new Error(
              'RoleType with Tag ' +
                v.t +
                ' is not present. Please regenerate your workspace.'
            );
          }

          const objectType = roleType.objectType;

          let role: unknown;

          if (objectType.isUnit) {
            role = unitFromJson(objectType.tag, v.v);
          } else {
            if (roleType.isOne) {
              role = v.o;
            } else {
              role = v.c;
            }
          }

          return [roleType, role];
        })
      );

      delete this.syncResponseRoles;
    }

    return this._roleByRoleType;
  }

  getRole(roleType: RoleType): unknown {
    return this.roleByRoleType?.get(roleType);
  }

  isPermitted(permission: number): boolean {
    if (permission == null) {
      return false;
    }

    if (this.grants == null) {
      return false;
    }

    if (
      this.revocations != null &&
      this.revocations.some((v) =>
        this.database.ranges.has(
          this.database.revocationById.get(v).permissionIds,
          permission
        )
      )
    ) {
      return false;
    }

    return this.grants.some((v) =>
      this.database.ranges.has(
        this.database.grantById.get(v).permissionIds,
        permission
      )
    );
  }
}
